typesafeconfig/typesafe_config.go

```diff
@@ -0,0 +1,340 @@
+
+	return &resolver{
+		baseNames:         []string{"application"},
+		configurationDirs: configurationDirs,
+		profiles:          profiles,
+	}
+}
+
+// Resolve given the provided options resolves your configuration
+func Resolve[T any](log *zap.SugaredLogger, options ...Option) (*T, error) {
+	r := defaultResolver()
+	for _, option := range options {
+		option(r)
+	}
+
+	if len(r.embeddedFilesystems) == 0 && len(r.configurationDirs) == 0 {
+		return nil, NoConfigurationSourcesProvided
+	}
+
+	candidates := getConfigurationFileCandidates(r.configurationDirs, r.baseNames, r.profiles)
+	sources, err := loadFileBasedConfigurationSources(log, candidates, r.embeddedFilesystems)
+	if err != nil {
+		return nil, err
+	}
+	sources = append(sources, loadEnvironmentSources())
+	untypedConfig := mergeSources(sources...)
+	if err = resolveSecrets(untypedConfig); err != nil {
+		return nil, err
+	}
+	var typeSafeConfig *T
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		WeaklyTypedInput: true,
+		Result:           &typeSafeConfig,
+		MatchName: func(mapKey, fieldName string) bool {
+			normalizedMapKey := strings.ToLower(mapKey)
+			normalizedMapKey = strings.ReplaceAll(normalizedMapKey, "-", "")
+			normalizedMapKey = strings.ReplaceAll(normalizedMapKey, "_", "")
+			return strings.ToLower(fieldName) == normalizedMapKey
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	err = decoder.Decode(untypedConfig)
+	return typeSafeConfig, err
+}
+
+func loadEnvironmentSources() map[string]interface{} {
+	config := make(map[string]interface{})
+	env := os.Environ()
+	for _, envVar := range env {
+		kvPair := strings.SplitN(envVar, "=", 2)
+		rawKey := kvPair[0]
+		value := kvPair[1]
+		key := strings.Split(rawKey, "_")
+		setValue(config, key, value)
+	}
+	return config
+}
+
+func setValue(config map[string]interface{}, key []string, value any) {
+	if len(key) == 1 {
+		config[key[0]] = value
+		return
+	}
+	cur, remaining := pop(key)
+	var nested map[string]interface{}
+	if config[cur] == nil {
+		nested = make(map[string]interface{})
+	} else {
+		curNested := config[cur]
+		unboxed, ok := curNested.(map[string]interface{})
+		if !ok {
+			nested = make(map[string]interface{})
+		} else {
+			nested = unboxed
+		}
+	}
+	config[cur] = nested
+	setValue(nested, remaining, value)
+}
+
+func pop[T any](array []T) (T, []T) {
+	return array[0], array[1:]
+}
+
+// mergeSources recursively left merges config sources, omitting any non-map values that are not one of: lists, numbers, or booleans
+// un-flattens keys before merging into new map
+func mergeSources(sources ...map[string]interface{}) map[string]interface{} {
+	m := make(map[string]interface{})
+	for _, unNormalizedSource := range sources {
+		source := normalizeKeys(unNormalizedSource)
+		// iterate through key and if the value is a map recurse, else set the key to the value if type is a number, list or boolean
+		for _, key := range maps.Keys(source) {
```

Comment:
`maps.Keys(...)` is a little strange here because

```
for key := range source {
...
}
```
will do the same thing

Reply:
Good to know, I'll fix this.